### PR TITLE
fix(Scripts/BlackTemple): Rewrite Shade of Akama

### PR DESCRIPTION
**Pull request by Gultask**
## Changes Proposed:

This PR proposes changes to:
-  [ ] Core (units, players, creatures, game systems).
-  [X] Scripts (bosses, spell scripts, creature scripts).
-  [X] Database (SAI, creatures, etc).

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/9095
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/18591

## SOURCE:

The changes have been validated through:
- [X] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [X] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [X] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1. Fight Shade of Akama
2. Go to Illidari Council and defeat them, see if Akama's event plays
3. Deafeat Illidari Council but, instead of starting the event, crash the server and return, see if the door to Illidan is open

## Known Issues and TODO List:

- [x] Illidari Council
- [x] Shade of Akama Generators
- [x] GetCreature not working

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by pklloveyou** [comment]
```
DELETE FROM `smart_scripts` WHERE `source_type` = 0 AND `entryorguid` = 23375;
UPDATE `creature_template` SET `AIName` = '', `ScriptName` = 'npc_shadow_demon' WHERE `entry` = 23375;
```
```
class npc_shadow_demon : public CreatureScript
{
public:
    npc_shadow_demon() : CreatureScript("npc_shadow_demon") { }

    struct npc_shadow_demonAI : public ScriptedAI
    {
        npc_shadow_demonAI(Creature* creature) : ScriptedAI(creature) { }

        void Reset() override
        {
            me->SetReactState(REACT_PASSIVE);
            me->CastSpell(me, SPELL_FIND_TARGET, true);
        }

        void SpellHitTarget(Unit* /*target*/, SpellInfo const* spell) override
        {
            if (spell->Id == SPELL_FIND_TARGET)
            {
                Unit* player = me->SelectNearestTarget(100.0f, true);
                if (player && player->IsAlive())
                {
                    PlayerGUID.Clear();
                    PlayerGUID = player->GetGUID();
                    me->CastSpell(player, 41083, false);
                    me->AddThreat(player, 1000000.0f);
                    me->GetMotionMaster()->MoveFollow(player, 1.0f, me->GetAngle(player));
                }
            }
        }

        void MovementInform(uint32 type, uint32 /*id*/) override
        {
            if (type != FOLLOW_MOTION_TYPE)
                return;
            me->InterruptNonMeleeSpells(false);
            if (Player* player = ObjectAccessor::GetPlayer(*me, PlayerGUID))
                me->CastSpell(player, 41082, true);
        }

    private:
        ObjectGuid PlayerGUID;
    };

    CreatureAI* GetAI(Creature* creature) const override
    {
        return GetBlackTempleAI<npc_shadow_demonAI>(creature);
    }
};
```

**Comment by pklloveyou** [comment]
```
DELETE FROM `smart_scripts` WHERE `source_type` = 0 AND `entryorguid` = 23375;
UPDATE `creature_template` SET `AIName` = '', `ScriptName` = 'npc_shadow_demon' WHERE `entry` = 23375;
```
```
class npc_shadow_demon : public CreatureScript
{
public:
    npc_shadow_demon() : CreatureScript("npc_shadow_demon") { }

    struct npc_shadow_demonAI : public ScriptedAI
    {
        npc_shadow_demonAI(Creature* creature) : ScriptedAI(creature) { }

        void Reset() override
        {
            me->SetReactState(REACT_PASSIVE);
            me->CastSpell(me, SPELL_FIND_TARGET, true);
        }

        void SpellHitTarget(Unit* /*target*/, SpellInfo const* spell) override
        {
            if (spell->Id == SPELL_FIND_TARGET)
            {
                Unit* player = me->SelectNearestTarget(100.0f, true);
                if (player && player->IsAlive())
                {
                    PlayerGUID.Clear();
                    PlayerGUID = player->GetGUID();
                    me->CastSpell(player, 41083, false);
                    me->AddThreat(player, 1000000.0f);
                    me->GetMotionMaster()->MoveFollow(player, 1.0f, me->GetAngle(player));
                }
            }
        }

        void MovementInform(uint32 type, uint32 /*id*/) override
        {
            if (type != FOLLOW_MOTION_TYPE)
                return;
            me->InterruptNonMeleeSpells(false);
            if (Player* player = ObjectAccessor::GetPlayer(*me, PlayerGUID))
                me->CastSpell(player, 41082, true);
        }

    private:
        EventMap events;
        ObjectGuid PlayerGUID;
    };

    CreatureAI* GetAI(Creature* creature) const override
    {
        return GetBlackTempleAI<npc_shadow_demonAI>(creature);
    }
};
```

**Comment by pklloveyou** [comment]
Maybe we still need this

**Comment by pklloveyou** [comment]
Maybe we still need this
image
shit!
My network is not good

File: src/server/scripts/Outland/BlackTemple/boss_illidan.cpp [hunk 1]
```diff
@@ -830,6 +842,22 @@ class npc_akama_illidan : public CreatureScript
...
+            {
+                ScheduleUniqueTimedEvent(200ms, [&]
+                {
+                    Talk(SAY_AKAMA_COUNCIL_1);
+                }, 1);
+                ScheduleUniqueTimedEvent(7800ms, [&]
+                {
+                    Talk(SAY_AKAMA_COUNCIL_2);
+                    me->SetNpcFlag(UNIT_NPC_FLAG_GOSSIP);
+                }, 2);
```
Comment (pklloveyou) [inline src/server/scripts/Outland/BlackTemple/boss_illidan.cpp:857]: 1 to 1s ?
2 to 2s ?

File: src/server/scripts/Outland/BlackTemple/illidari_council.cpp [hunk 2]
```diff
@@ -296,14 +293,14 @@ class boss_gathios_the_shatterer : public CreatureScript
         Creature* SelectCouncilMember()
         {
             if (roll_chance_i(50))
-                return ObjectAccessor::GetCreature(*me, instance->GetGuidData(NPC_LADY_MALANDE));
+                instance->GetCreature(DATA_LADY_MALANDE);
```
Comment (pklloveyou) [inline src/server/scripts/Outland/BlackTemple/illidari_council.cpp:296]: return instance->GetCreature(DATA_LADY_MALANDE);

File: src/server/scripts/Outland/BlackTemple/instance_black_temple.cpp [hunk 3]
```diff
@@ -78,138 +104,56 @@ class instance_black_temple : public InstanceMapScript
...
+                case NPC_ASHTONGUE_STORMCALLER:
+                case NPC_ASHTONGUE_PRIMALIST:
+                case NPC_ASHTONGUE_FERAL_SPIRIT:
+                case NPC_ASHTONGUE_STALKER:
+                case NPC_STORM_FURY:
+                    if (GetBossState(DATA_SHADE_OF_AKAMA) == DONE)
+                        creature->SetFaction(FACTION_ASHTONGUE_DEATHSWORN);
+                    else
+                        ashtongueGUIDs.insert(creature->GetGUID());
                     break;
```
Comment (pklloveyou) [inline src/server/scripts/Outland/BlackTemple/instance_black_temple.cpp:138]: default:
                break;

File: data/sql/updates/pending_db_world/rev_1711048056647652400.sql [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+(23089, 9, 0, 'Those who\'ve defiled this temple have all been defeated.  All but one!', 12, 0, 100, 1, 0, 0, 21518, 0, 'SAY_AKAMA_COUNCIL_1'),
+(23089, 10, 0, 'Let us finish what we\'ve started.  I will lead you to Illidan\'s abode once you\'ve recovered your strength.', 12, 0, 100, 1, 0, 0, 21520, 0, 'SAY_AKAMA_COUNCIL_2');
+
+DELETE FROM `waypoint_data` WHERE `id` = 230891;
+INSERT INTO `waypoint_data` (`id`, `point`, `position_x`, `position_y`, `position_z`, `move_type`) VALUES
+(230891, 1, 642.5905, 305.6287, 271.6885, 1),
+(230891, 2, 660.76465, 305.76627, 271.70413, 1);
+
+UPDATE `creature` SET `phaseMask` = 3, WHERE guid = 148734;
+UPDATE `creature` SET `phaseMask` = 5, WHERE guid = 148733;
```
Comment (pklloveyou) [inline data/sql/updates/pending_db_world/rev_1711048056647652400.sql:43]: -,
-,
```
UPDATE `creature` SET `phaseMask` = 3 WHERE guid = 148734;
UPDATE `creature` SET `phaseMask` = 5 WHERE guid = 148733;
```

File: src/server/scripts/Outland/BlackTemple/illidari_council.cpp [hunk 5]
```diff
@@ -152,14 +152,13 @@ class boss_illidari_council : public CreatureScript
...
+        Creature* councilGUIDs[4];
 
         void Reset() override
         {
             BossAI::Reset();
-            Creature* member = nullptr;
             for (uint8 i = 0; i < 4; ++i)
-                if ((member = ObjectAccessor::GetCreature(*me, councilGUIDs[i])))
+                if (Creature* member = councilGUIDs[i])
                     member->AI()->EnterEvadeMode();
```
Comment (pklloveyou) [inline src/server/scripts/Outland/BlackTemple/illidari_council.cpp:162]: I'm crash here

**Comment by pklloveyou** [comment]
```
DELETE FROM `smart_scripts` WHERE `source_type` = 2 AND `entryorguid` = 4665;
INSERT INTO `smart_scripts` VALUES (4665, 2, 0, 0, 46, 0, 100, 1, 0, 0, 0, 0, 0, 0, 45, 0, 1, 0, 0, 0, 0, 10, 148358, 22871, 0, 0, 0, 0, 0, 0, 'Area Trigger - On Trigger - Talk');

```

**Comment by Gultask** [comment]
thank you pk! 😁

**Comment by heyitsbench** [comment]
At last, a ✔️.

**Comment by Gultask** [comment]
I almost had one before I went to sleep

**Comment by Gultask** [comment]
@pklloveyou I think the crash is due to councilGUIDs being initialized with some trash, because it only *sometimes* crashes

File: data/sql/updates/pending_db_world/rev_1711048056647652400.sql [hunk 6]
```diff
@@ -39,5 +39,5 @@ INSERT INTO `waypoint_data` (`id`, `point`, `position_x`, `position_y`, `positio
 (230891, 1, 642.5905, 305.6287, 271.6885, 1),
 (230891, 2, 660.76465, 305.76627, 271.70413, 1);
 
-UPDATE `creature` SET `phaseMask` = 3, WHERE guid = 148734;
-UPDATE `creature` SET `phaseMask` = 5, WHERE guid = 148733;
\ No newline at end of file
+UPDATE `creature` SET `phaseMask` = 3, WHERE `guid` = 148734;
+UPDATE `creature` SET `phaseMask` = 5, WHERE `guid` = 148733;
```
Comment (pklloveyou) [inline data/sql/updates/pending_db_world/rev_1711048056647652400.sql]: ,

**Comment by pklloveyou** [comment]
@Gultask ObjectGuid is safer than Creature*

**Comment by pklloveyou** [comment]
image
image
Here's how I did it.

File: src/server/scripts/Outland/BlackTemple/boss_illidan.cpp [hunk 7]
```diff
@@ -913,8 +942,8 @@ class npc_akama_illidan : public CreatureScript
                         udalo->CastSpell(udalo, SPELL_AKAMA_DOOR_OPEN, false);
                     break;
                 case EVENT_AKAMA_SCENE_9:
-                    instance->SetBossState(DATA_AKAMA_FINISHED, NOT_STARTED);
-                    instance->SetBossState(DATA_AKAMA_FINISHED, DONE);
+                    instance->SetBossState(DATA_AKAMA_ILLIDAN, NOT_STARTED);
+                    instance->SetBossState(DATA_AKAMA_ILLIDAN, DONE);
                     break;
                 case EVENT_AKAMA_SCENE_10:
                     Talk(SAY_AKAMA_BEWARE);
```
Comment (pklloveyou) [inline src/server/scripts/Outland/BlackTemple/boss_illidan.cpp:949]: - if (Creature* illidan = ObjectAccessor::GetCreature(*me, instance->GetGuidData(NPC_ILLIDAN_STORMRAGE)))
+ if (Creature* illidan = instance->GetCreature(DATA_ILLIDAN_STORMRAGE))

File: src/server/scripts/Outland/BlackTemple/boss_shade_of_akama.cpp [hunk 8]
```diff
@@ -25,251 +25,175 @@ enum Says
...
+    ACTION_GENERATOR_START          = 1,
+    ACTION_GENERATOR_STOP           = 2,
+    ACTION_GENERATOR_DESPAWN_ALL    = 3,
+
+    COUNTER_SPAWNS_MAX              = 20,   // Max number of spawns for each generator, number chosen at random
+
+    ACTION_AKAMA_START_OUTRO        = 1,
+
+    FACTION_DEFAULT                 = 1820,
+    FACTION_ENGAGE                  = 1868
```
Comment (pklloveyou) [inline src/server/scripts/Outland/BlackTemple/boss_shade_of_akama.cpp:84]: 1866？

File: src/server/scripts/Outland/BlackTemple/boss_shade_of_akama.cpp [hunk 9]
```diff
@@ -25,251 +25,175 @@ enum Says
...
-        me->SetReactState(REACT_PASSIVE);
         me->SetUnitFlag(UNIT_FLAG_NOT_SELECTABLE);
-        me->SetImmuneToAll(true);
         me->SetWalk(true);
+        me->SetReactState(REACT_DEFENSIVE);
+        BossAI::Reset();
     }
 
     void EnterEvadeMode(EvadeReason why) override
     {
```
Comment (pklloveyou) [inline src/server/scripts/Outland/BlackTemple/boss_shade_of_akama.cpp:107]: me->GetCreatureListWithEntryInGrid(channelers, NPC_ASHTONGUE_CHANNELER, 40.0f);
            for (auto const& channeler : channelers)
                channeler->SetUnitFlag(UNIT_FLAG_NOT_SELECTABLE);

File: src/server/scripts/Outland/BlackTemple/illidari_council.cpp [hunk 10]
```diff
@@ -152,14 +152,13 @@ class boss_illidari_council : public CreatureScript
...
         }
 
-        ObjectGuid councilGUIDs[4];
+        Creature* councilGUIDs[4] = { nullptr };
 
         void Reset() override
         {
             BossAI::Reset();
-            Creature* member = nullptr;
             for (uint8 i = 0; i < 4; ++i)
```
Comment (elthehablo) [inline src/server/scripts/Outland/BlackTemple/illidari_council.cpp:160]:
```suggestion
            for (uint8 DATA_GATHIOS_THE_SHATTERER; i <= DATA_VERAS_DARKSHADOW; ++i)
```

File: src/server/scripts/Outland/BlackTemple/illidari_council.cpp [hunk 11]
```diff
@@ -152,14 +152,13 @@ class boss_illidari_council : public CreatureScript
...
-                if ((member = ObjectAccessor::GetCreature(*me, councilGUIDs[i])))
+                if (Creature* member = councilGUIDs[i])
```
Comment (elthehablo) [inline src/server/scripts/Outland/BlackTemple/illidari_council.cpp:161]:
```suggestion
                if (Creature* member = instance->GetCreature(i))
```

File: src/server/scripts/Outland/BlackTemple/boss_shade_of_akama.cpp (hunk 9 above)
Comment (elthehablo) [reply src/server/scripts/Outland/BlackTemple/boss_shade_of_akama.cpp:107]: Maybe we can make them minions and then you can do for all minions. But tbh idk how this fight is and if they're suitable as minions. See Magtheridon for examples.

File: src/server/scripts/Outland/BlackTemple/boss_shade_of_akama.cpp (hunk 8 above)
Comment (Gultask) [reply src/server/scripts/Outland/BlackTemple/boss_shade_of_akama.cpp:84]: 1868 is sniffed. Shade of Akama itself has a Monster Spar faction
```
ServerToClient: SMSG_UPDATE_OBJECT (0x27C3) Length: 92 ConnIdx: 1 Time: 11/25/2022 17:57:10.681 Number: 52899
NumObjUpdates: 1
MapID: 564 (564)
HasRemovedObjects: False
Data size: 81
[0] UpdateType: Values
[0] Object Guid: Full: \\\\ Creature/0 R3881/S31776 Map: 564 Entry: 23191 Low: 75740
[0] [0] ChannelObjects: Full: \\\\ Creature/0 R3881/S31776 Map: 564 Entry: 22841 Low: 75740
[0] (ChannelData) SpellID: 40447
[0] (ChannelData) (SpellVisual) SpellXSpellVisualID: 18249
[0] (ChannelData) (SpellVisual) ScriptVisualID: 0
[0] FactionTemplate: 1868
[0] Flags: 262144
[0] [0] NpcFlags: 0
```

File: src/server/scripts/Outland/BlackTemple/boss_shade_of_akama.cpp (hunk 9 above)
Comment (Gultask) [reply src/server/scripts/Outland/BlackTemple/boss_shade_of_akama.cpp:107]: Will add it to OnEvade instead

File: src/server/scripts/Outland/BlackTemple/illidari_council.cpp (hunk 10 above)
Comment (Gultask) [reply src/server/scripts/Outland/BlackTemple/illidari_council.cpp:160]: don't we need
```
uint8 i = DATA_GATHIOS_THE_SHATTERER
```
?

File: src/server/scripts/Outland/BlackTemple/boss_shade_of_akama.cpp [hunk 12]
```diff
@@ -279,242 +203,257 @@ struct npc_akama_shade : public ScriptedAI
...
-        summons.Summon(summon);
+        CloseGossipMenuFor(player);
+        me->RemoveNpcFlag(UNIT_NPC_FLAG_GOSSIP);
+        me->RemoveAurasDueToSpell(SPELL_STEALTH);
+        me->GetMotionMaster()->MovePoint(POINT_ENGAGE, AkamaEngage);
     }
 
     void UpdateAI(uint32 diff) override
     {
-        events2.Update(diff);
```
[the remark this answers is not in the document]
Comment (Gultask) [reply src/server/scripts/Outland/BlackTemple/boss_shade_of_akama.cpp:331]: hmm?

File: src/server/scripts/Outland/BlackTemple/illidari_council.cpp (hunk 10 above)
Comment (elthehablo) [reply src/server/scripts/Outland/BlackTemple/illidari_council.cpp:160]: yeah ofc sorry I typod

File: src/server/scripts/Outland/BlackTemple/illidari_council.cpp (hunk 11 above)
Comment (pklloveyou) [reply src/server/scripts/Outland/BlackTemple/illidari_council.cpp:161]:
```
for (uint32 i = DATA_GATHIOS_THE_SHATTERER; i < DATA_VERAS_DARKSHADOW + 1; ++i)
    if (Creature* member = instance->GetCreature(i))
```
this?

**Comment by pklloveyou** [comment]
or i <= ?

**Comment by elthehablo** [comment]
> or i <= ?

This, because that + 1 is ugly imo

**Comment by Gultask** [comment]
I think it's fine now

File: src/server/scripts/Outland/BlackTemple/boss_shade_of_akama.cpp [hunk 13]
```diff
@@ -25,251 +25,178 @@ enum Says
...
-                    summon->RemoveUnitFlag(UNIT_FLAG_NOT_SELECTABLE);
-                }
-                else if (action == ACTION_KILL_CHANNELERS)
-                {
-                    Unit::Kill(me, summon);
-                }
-            }
-    }
+    std::list<Creature*> channelers;
+    std::list<Creature*> generators;
```
Comment (elthehablo) [inline src/server/scripts/Outland/BlackTemple/boss_shade_of_akama.cpp:96]: best to clear this on reset

Comment (Gultask) [reply src/server/scripts/Outland/BlackTemple/boss_shade_of_akama.cpp:96]: real

**Review by elthehablo — APPROVED** [review]

**Comment by icaca** [comment]
After losing the battle with Illidan, Akama did not return to his original position and could not talk to him.

**Comment by elthehablo** [comment]
> After losing the battle with Illidan, Akama did not return to his original position and could not talk to him.

Then make a bug report please

**Pull request merged**